File: src/main/java/org/cimsbioko/navconfig/NavigatorConfig.java
package org.cimsbioko.navconfig;

import android.util.Log;
import org.cimsbioko.navconfig.forms.Binding;
import org.cimsbioko.scripting.JsConfig;
import java.util.*;

import static java.util.Collections.*;
import static org.cimsbioko.utilities.CampaignUtils.loadCampaign;


/**
 * The configuration source for hierarchy navigation, form display and form data binding for the field worker section of
 * the application.
 *
 * @see org.cimsbioko.activity.FieldWorkerActivity
 * @see org.cimsbioko.activity.HierarchyNavigatorActivity
 */
public class NavigatorConfig {

    private static final String TAG = NavigatorConfig.class.getSimpleName();

    private static NavigatorConfig instance;

    private Hierarchy hierarchy;
    private Map<String, NavigatorModule> modules = emptyMap();
    private Map<String, Binding> bindings = emptyMap();
    private JsConfig config;


    protected NavigatorConfig() {
        init();
    }

    public Hierarchy getHierarchy() {
        return hierarchy;
    }

    private void init() {
        loadConfig();
        consolidateFormBindings();
    }

    /**
     * Reloads the configuration, ensuring url resource caching doesn't get in the way.
     */
    public void reload() {
        config.close();
        init();
    }

    /*
     * Define the navigation modules. They will show up in the interface in the order specified.
     */
    private void loadConfig() {
        modules = new LinkedHashMap<>();
        try {
            config = loadCampaign();
            hierarchy = config.getHierarchy();
            for (NavigatorModule module : config.getNavigatorModules()) {
                modules.put(module.getName(), module);
            }
        } catch (Exception e) {
            Log.e(TAG, "failure initializing js config", e);
        }
    }

    /**
     * Creates an index of all module {@link Binding} objects by name for fast access.
     */
    private void consolidateFormBindings() {
        bindings = new HashMap<>();
        for (NavigatorModule module : modules.values()) {
            bindings.putAll(module.getBindings());
        }
    }

    public static synchronized NavigatorConfig getInstance() {
        if (instance == null) {
            instance = new NavigatorConfig();
        }
        return instance;
    }

    /**
     * Gets all configured navigator modules.
     *
     * @return a list of configured {@link NavigatorModule}s in definition order
     */
    public Collection<NavigatorModule> getModules() {
        return unmodifiableCollection(modules.values());
    }

    /**
     * Gets all configured navigator module names.
     *
     * @return a list of configured {@link NavigatorModule} names in definition order
     */
    public Set<String> getModuleNames() {
        return unmodifiableSet(modules.keySet());
    }

    /**
     * Get logical names for the configured hierarchy levels.
     *
     * @return a list of configured hier levels, from highest to lowest.
     */
    public List<String> getLevels() {
        return hierarchy.getLevels();
    }

    /**
     * Get logical names for admin hierarchy levels.
     *
     * @return a list of hier levels corresponding to admin hierarchy, from highest to lowest.
     */
    public List<String> getAdminLevels() {
        return hierarchy.getAdminLevels();
    }

    /**
     * Returns the level value corresponding to immediately more general value in the hierarchy.
     * For example, if 'subdistrict' was specified, 'district' would be the result.
     *
     * @param level the cimsbioko app level value (the child)
     * @return the level value corresponding to the parent level of the specified one
     */
    public String getParentLevel(String level) {
        return hierarchy.getParentLevel(level);
    }

    /**
     * Convenience method for getting top-most level. Same as getLevels().get(0).
     *
     * @return the top-most level
     */
    public String getTopLevel() {
        return getLevels().get(0);
    }

    /**
     * Convenience method for getting bottom-most level.
     *
     * @return the bottom-most level
     */
    public String getBottomLevel() {
        return getLevels().get(getLevels().size() - 1);
    }

    /**
     * Get a localized label for the logical hierarchy level.
     *
     * @param level the hierarchy level
     * @return the level's label for the current locale
     */
    public String getLevelLabel(String level) {
        return getString(hierarchy.getLevelLabels().get(level));
    }

    public NavigatorModule getModule(String name) {
        return modules.get(name);
    }

    /**
     * Finds a configured form binding by name.
     *
     * @param name the name of the binding to retrieve
     * @return the {@link Binding} object for the given name or null if none exists
     */
    public Binding getBinding(String name) {
        return bindings.get(name);
    }

    /**
     * Get a localized string from the {@link JsConfig}'s {@link java.util.ResourceBundle}.
     *
     * @param key the resource key for a localized string
     * @return the string, localized for the current {@link java.util.Locale}
     */
    public String getString(String key) {
        return config.getString(key);
    }
}

File: src/main/kotlin/org/cimsbioko/navconfig/NavigatorConfig.kt
package org.cimsbioko.navconfig

import android.util.Log
import org.cimsbioko.scripting.JsConfig
import org.cimsbioko.search.SearchQueryBuilder
import org.cimsbioko.search.SearchSource
import org.cimsbioko.utilities.CampaignUtils.loadCampaign

/**
 * The configuration source for hierarchy navigation, form display and form data binding for the field worker section of
 * the application.
 *
 * @see org.cimsbioko.activity.FieldWorkerActivity
 * @see org.cimsbioko.activity.HierarchyNavigatorActivity
 */
class NavigatorConfig private constructor() {

    init {
        init()
    }

    private fun init() {
        loadConfig()
        consolidateFormBindings()
    }

    private lateinit var hierarchy: Hierarchy
    private lateinit var mods: Map<String, NavigatorModule>
    private lateinit var bindings: Map<String, Binding>
    private lateinit var config: JsConfig

    var adminSecret: String? = null
        private set

    lateinit var searchSources: Map<String, SearchSource>
        private set

    lateinit var searchQueryBuilder: SearchQueryBuilder

    /**
     * Reloads the configuration, ensuring url resource caching doesn't get in the way.
     */
    fun reload() {
        config.close()
        init()
    }

    /*
     * Define the navigation modules. They will show up in the interface in the order specified.
     */
    private fun loadConfig() {
        try {
            config = loadCampaign()
            hierarchy = config.hierarchy
            adminSecret = config.adminSecret
            mods = config.navigatorModules.associateBy { it.name }
            searchSources = config.searchSources
            searchQueryBuilder = config.searchQueryBuilder
        } catch (e: Exception) {
            Log.e(TAG, "failure initializing js config", e)
        }
    }

    /**
     * Creates an index of all module [Binding] objects by name for fast access.
     */
    private fun consolidateFormBindings() {
        bindings = HashMap<String, Binding>().apply { mods.values.forEach { putAll(it.bindings) } }
    }

    /**
     * Gets all configured navigator modules.
     *
     * @return a list of configured [NavigatorModule]s in definition order
     */
    val modules: Collection<NavigatorModule>
        get() = mods.values

    /**
     * Gets all configured navigator module names.
     *
     * @return a list of configured [NavigatorModule] names in definition order
     */
    val moduleNames: Set<String>
        get() = mods.keys

    /**
     * Get logical names for the configured hierarchy levels.
     *
     * @return a list of configured hier levels, from highest to lowest.
     */
    val levels: List<String>
        get() = hierarchy.levels

    /**
     * Get logical names for admin hierarchy levels.
     *
     * @return a list of hier levels corresponding to admin hierarchy, from highest to lowest.
     */
    val adminLevels: List<String>
        get() = hierarchy.adminLevels

    /**
     * Returns the level value corresponding to immediately more general value in the hierarchy.
     * For example, if 'subdistrict' was specified, 'district' would be the result.
     *
     * @param level the cimsbioko app level value (the child)
     * @return the level value corresponding to the parent level of the specified one
     */
    fun getParentLevel(level: String): String? = hierarchy.getParentLevel(level)

    /**
     * Convenience method for getting top-most level. Same as getLevels().get(0).
     *
     * @return the top-most level
     */
    val topLevel: String?
        get() = levels.firstOrNull()

    /**
     * Convenience method for getting bottom-most level.
     *
     * @return the bottom-most level
     */
    val bottomLevel: String?
        get() = levels.lastIndex.let { if (it in levels.indices) levels[it] else null }

    /**
     * Get a localized label for the logical hierarchy level.
     *
     * @param level the hierarchy level
     * @return the level's label for the current locale
     */
    fun getLevelLabel(level: String): String? = getString(hierarchy.levelLabels[level])

    fun getModule(name: String): NavigatorModule? = mods[name]

    /**
     * Finds a configured form binding by name.
     *
     * @param name the name of the binding to retrieve
     * @return the [Binding] object for the given name or null if none exists
     */
    fun getBinding(name: String): Binding? = bindings[name]

    /**
     * Returns the unique set of form ids bound with the loaded [Binding]s.
     *
     * @return  set of form ids from the loaded bindings
     */
    val formIds: Set<String>
        get() = HashSet<String>().apply { bindings.values.forEach { add(it.form) } }

    /**
     * Get a localized string from the [JsConfig]'s [java.util.ResourceBundle].
     *
     * @param key the resource key for a localized string
     * @return the string, localized for the current [java.util.Locale]
     */
    fun getString(key: String?): String = config.getString(key)

    companion object {
        private val TAG = NavigatorConfig::class.java.simpleName
        val instance: NavigatorConfig by lazy { NavigatorConfig() }
    }
}